### ISSUE TITLE
Build should statically detect when an invisible unicode character, such as U+200B (zero width space), sneak into our sources

### ISSUE THREAD

**Issue by mikemccand** [apache/lucene#12931]
### Description

Spinoff from #12911 where I hit weird/scary exceptions during `./gradlew regenerate` with certain JDK version(s) and not others due to a lurking zero width space Unicode character.  @uschindler fixed that one case, but perhaps we have others?  And perhaps copy/apply of a patch file is another vector for them to invade us.

Can we statically assert we don't accidentally invite these little buggers in?

**Comment by rmuir** [comment]
But the problem there wasn't the ZWSP but wrong character encoding? It also would have happened with a smiley emoji or arabic letter?

**Comment by rmuir** [comment]
Also lots of tests have strings that have "non-visible" unicode characters. I think this issue is too broad and will never work.

Instead I think we should limit variable, method, class names, etc to simple ascii? But lets not police what is inside test strings please.

**Comment by dweiss** [comment]
Wouldn't it be better if they were explicitly (\uxxxx) encoded though? Especially those invisible ones, which can be really annoying since you can't see them...

**Comment by uschindler** [comment]
> But the problem there wasn't the ZWSP but wrong character encoding? It also would have happened with a smiley emoji or arabic letter?

Yes. The problem was still that if you do not see it, you can't fix it. Not even the text editor showed them. It took me long time to fix this issue. Yes it was a default charset problem on Mike's device.

I generally agree with Mike on some of his points: if you need to write a test that requires some strange invisible character, please add it as Unicode escape in java code.

So I would like to add a check/pattern to or source patterns task that complains about hidden code. If there's a test that gets complaints: oh man just add an escape and all are happy.

**Comment by rmuir** [comment]
> Wouldn't it be better if they were explicitly (\uxxxx) encoded though? Especially those invisible ones, which can be really annoying since you can't see them...

But how in the world are you supposed to read such text?

**Comment by rmuir** [comment]
I'm looking into this one: https://errorprone.info/bugpattern/UnicodeInCode

This wouldn't cause any problems for comments or literals which is my only concern: but would prevent crazy stuff in code itself. If it works without zillions of failures I'll make a PR.

**Comment by rmuir** [comment]
OK see PR, this check just wasn't enabled because there was stuff to fix. I fixed the spatial3d code so now we can prevent this going forwards.

**Comment by uschindler** [comment]
> > Wouldn't it be better if they were explicitly (\uxxxx) encoded though? Especially those invisible ones, which can be really annoying since you can't see them...
> 
> But how in the world are you supposed to read such text?

Your PR only fixed the identifiers. I am solely talking about Java files. If you need a text string to pass to an analyzer that splits on whitespace whatever, why can't we forbid invisible code points. Just use a escaped text string.

I am not talking about resource files. Just java code. And there I would like to enforce all string declarations with such characters to use Unicode escapes.

**Comment by uschindler** [comment]
...just the hidden ones.

I agree with your PR for java code, as this is the security issue for patches coming in. 👍

P.S. the PR wouldn't have prevented the original issue by Mike, because this was Gradle code not passing through javac, but directly interpreted.

**Comment by dweiss** [comment]
> If you need a text string to pass to an analyzer that splits on whitespace whatever, why can't we forbid invisible code points. Just use a escaped text string.

I'm not against emojis in identifiers :). But I had to debug what Uwe mentioned more than a few times in my life - not just in source but also in data (jsons, etc.) - takes a while to figure out what's wrong when you can't see it.

**Comment by uschindler** [comment]
Hi,
here is my simplest one to just prevent 2 unicode characters: #12937

In `validateSourcePatterns` we have more of those. It detected also the violation I introduced yesterday. The problem with u200B is mainly: Javadocs generated by the javadoc tool often separate method names from their parameters using that codepoint. If you copy/paste a method declaration, you insert it into code. This would possibly also be detected by Robert's errorprone, but this one applies to all source files.

**Comment by rmuir** [comment]
i'm ok with the ZWSP and BOM marker. just not other characters that one might consider invisible such as ZWJ. i want the tests to use human readable strings. Sorry, I'll be sticking to my guns on that.

**Comment by uschindler** [comment]
I will merge the other PR soon, will just make reporting better. The other PR also detects invalid UTF-8.

**Comment by msokolov** [comment]
What about ZWNJ? TBH I don't understand the difference between ZWS and ZWNJ

**Comment by uschindler** [comment]
> What about ZWNJ? TBH I don't understand the difference between ZWS and ZWNJ

I could add it to the regex, but then Robert will shot me with his guns...

**Comment by uschindler** [comment]
To give a bit of context: I would also add ZWNJ (and if I do, 3 tests trigger the validation violation). But Robert does not like it because he wants to write tests without escapes. I don't agree with that, but I don't want to argue.

To me the more important one is ZWSP (U+200B), because when you copy-paste code from javadocs generated by the javadoc tool, there are U+200B between method names and parameters, so it happens easily that you get them into code. This happened to me in the `ExtractJdkApis.java` file used by Gradle to generate API signature files.

So my PR #12937 just catches this one for now. In addition I added validation of the UTF-8 encoding, so files saved accidentally as ISO-8859-x are detected by `./gradlew validateSourcePatterns`.

**Comment by mikemccand** [comment]
Thanks @uschindler and @rmuir!
